### DISCUSSION TITLE
jq remove empty array and empty entries

### DISCUSSION THREAD

**Discussion by lamyergeier** [stedolan/jq#3170] [Q&A]
In my object array I have some object that have empty entries and empty array as follows

```
[
  {
    "dTarget": null,
    "SourceP": [
      null
    ]
  }
]
```

How can I remove all such empty entries, arrays, and empty objects?

**Comment by ShreeChaturvedi** [comment]
`walk` (jq 1.6+) hits nested values first, so `[null]` turns into `[]` and that empty array gets dropped too. I'd skip `del(..|nulls)` here, it leaves `"SourceP": []`. Your sample comes out `[]`.

```jq
walk(
  if type == "object" then
    with_entries(select(.value != null and .value != [] and .value != {}))
  elif type == "array" then
    map(select(. != null and . != [] and . != {}))
  else .
  end
)
```
